sdks/python/apache_beam/ml/transforms/handlers.py

```diff
@@ -0,0 +1,406 @@
+    """
+    Return a DatasetMetadata object to be used with
+    tft_beam.AnalyzeAndTransformDataset.
+    Args:
+      input_types: A dictionary of column names and types.
+    Returns:
+      A DatasetMetadata object.
+    """
+    raw_data_feature_spec = {}
+    for key, value in input_types.items():
+      raw_data_feature_spec[key] = self._get_raw_data_feature_spec_per_column(
+          typ=value, col_name=key)
+    raw_data_metadata = dataset_metadata.DatasetMetadata(
+        schema_utils.schema_from_feature_spec(raw_data_feature_spec))
+    return raw_data_metadata
+
+  def _get_raw_data_feature_spec_per_column(self, typ: type, col_name: str):
+    """
+    Return a FeatureSpec object to be used with
+    tft_beam.AnalyzeAndTransformDataset
+    Args:
+      typ: A type of the column.
+      col_name: A name of the column.
+    Returns:
+      A FeatureSpec object.
+    """
+    # lets conver the builtin types to typing types for consistency.
+    typ = native_type_compatibility.convert_builtin_to_typing(typ)
+    containers_type = (List._name, Tuple._name)
+    is_container = hasattr(typ, '_name') and typ._name in containers_type
+
+    if is_container:
+      dtype = typing.get_args(typ)[0]
+      if len(typing.get_args(typ)) > 1 or typing.get_origin(dtype) == Union:
+        raise RuntimeError(
+            f"Incorrect type specifications in {typ} for column {col_name}. "
+            f"Please specify a single type.")
+      if dtype not in _default_type_to_tensor_type_map:
+        raise TypeError(
+            f"Unable to identify type: {dtype} specified on column: {col_name}"
+            f". Please specify a valid type.")
+    else:
+      dtype = typ
+
+    is_container = is_container or issubclass(dtype, np.generic)
+    if is_container:
+      return tf.io.VarLenFeature(_default_type_to_tensor_type_map[dtype])
+    else:
+      return tf.io.FixedLenFeature([], _default_type_to_tensor_type_map[dtype])
+
+  def get_metadata(self, input_types: Dict[str, type]):
+    """
+    Return metadata to be used with tft_beam.AnalyzeAndTransformDataset
+    Args:
+      input_types: A dictionary of column names and types.
+    """
+    raise NotImplementedError
+
+  def write_transform_artifacts(self, transform_fn, location):
+    """
+    Write transform artifacts to the given location.
+    Args:
+      transform_fn: A transform_fn object.
+      location: A location to write the artifacts.
+    Returns:
+      A PCollection of WriteTransformFn writing a TF transform graph.
+    """
+    return (
+        transform_fn
+        | 'Write Transform Artifacts' >>
+        transform_fn_io.WriteTransformFn(location))
+
+  def infer_output_type(self, input_type):
+    if not isinstance(input_type, RowTypeConstraint):
+      row_type = RowTypeConstraint.from_user_type(input_type)
+    fields = row_type._inner_types()
+    return Dict[str, Union[tuple(fields)]]
+
+  def _get_artifact_location(self, pipeline: beam.Pipeline):
```

Comment:
Instead of dumping this in the staging directory or a temp directory, should we require users to provide an output directory? Presumably, users will want a well defined location for retrieving their artifacts

Reply:
Added it as optional. If user doesn't provide it, I am falling back to this approach

Reply:
I don't think the staging/temp directory makes sense as a default--if there are artifacts to be produced/consumed, this should be a required argument.